File: src/utils/mp_crypto.erl
-module(mp_crypto).
-export([derive_key/1,
         encrypt/2,
         decrypt/2]).

-define(IV_LENGTH, 12).
-define(TAG_LENGTH, 16).
-define(KEY_LENGTH, 16).
-define(PBKDF2_ITERATIONS, 10000).
-define(SALT, <<"make-proxy">>).

%% Derive a 128-bit encryption key from the configured password.
-spec derive_key(string() | binary()) -> binary().
derive_key(Password) when is_list(Password) ->
    derive_key(list_to_binary(Password));
derive_key(Password) when is_binary(Password) ->
    crypto:pbkdf2_hmac(sha256, Password, ?SALT, ?PBKDF2_ITERATIONS, ?KEY_LENGTH).

%% AES-128-GCM: random IV per message, ciphertext is authenticated.
%% Output layout: <<IV:12, Tag:16, CipherText/binary>>
-spec encrypt(binary(), binary()) -> binary().
encrypt(Key, Binary) ->
    IV = crypto:strong_rand_bytes(?IV_LENGTH),
    {CipherText, Tag} = crypto:crypto_one_time_aead(
        aes_128_gcm, Key, IV, Binary, <<>>, ?TAG_LENGTH, true),
    <<IV/binary, Tag/binary, CipherText/binary>>.

-spec decrypt(binary(), binary()) -> {ok, binary()} |
                                     {error, term()}.
decrypt(Key, <<IV:?IV_LENGTH/binary, Tag:?TAG_LENGTH/binary, CipherText/binary>>) ->
    case crypto:crypto_one_time_aead(aes_128_gcm, Key, IV, CipherText, <<>>, Tag, false) of
        error ->
            {error, decrypt_failed};
        PlainText ->
            {ok, PlainText}
    end;

decrypt(_, _) ->
    {error, invalid_data}.

File: android/app/src/main/java/com/ruoyudai/makeproxy/MpCrypto.kt
package com.ruoyudai.makeproxy

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-128-GCM with per-message random IV, mirroring the Erlang mp_crypto module.
 * Wire layout: <<IV:12, Tag:16, CipherText>>.
 * Key derived from password with PBKDF2-HMAC-SHA256, 10000 iterations.
 * Note: use an ASCII password - Erlang list_to_binary and Java PBEKeySpec
 * treat non-ASCII characters differently.
 */
object MpCrypto {
    private const val IV_LEN = 12
    private const val TAG_LEN = 16
    private const val ITERATIONS = 10000
    private val SALT = "make-proxy".toByteArray(Charsets.UTF_8)
    private val random = SecureRandom()

    fun deriveKey(password: String): ByteArray {
        val spec = PBEKeySpec(password.toCharArray(), SALT, ITERATIONS, 128)
        return SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
            .generateSecret(spec).encoded
    }

    fun encrypt(key: ByteArray, plain: ByteArray): ByteArray {
        val iv = ByteArray(IV_LEN).also { random.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"),
            GCMParameterSpec(TAG_LEN * 8, iv)
        )
        val ctAndTag = cipher.doFinal(plain) // Java appends tag at the end
        val ct = ctAndTag.copyOfRange(0, ctAndTag.size - TAG_LEN)
        val tag = ctAndTag.copyOfRange(ctAndTag.size - TAG_LEN, ctAndTag.size)
        return iv + tag + ct
    }

    fun decrypt(key: ByteArray, data: ByteArray): ByteArray {
        require(data.size >= IV_LEN + TAG_LEN) { "invalid data" }
        val iv = data.copyOfRange(0, IV_LEN)
        val tag = data.copyOfRange(IV_LEN, IV_LEN + TAG_LEN)
        val ct = data.copyOfRange(IV_LEN + TAG_LEN, data.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"),
            GCMParameterSpec(TAG_LEN * 8, iv)
        )
        return cipher.doFinal(ct + tag)
    }
}
